spec/consensus/light-client.md

```diff
@@ -60,259 +61,368 @@ For the purpose of this lite client specification, we assume that the Tendermint
+**Observation 1.** If *h.Header.bfttime + tp > now*, we trust the old
+validator set *h.Header.NextV*.
 
-**Observation 1.** If _h.Header.bfttime + tp > now_, we trust the old
-validator set _h.Header.NextV_.
+When we say we trust *h.Header.NextV* we do *not* trust that each individual validator in *h.Header.NextV* is correct, but we only trust the fact that at most 1/3 of them are faulty (more precisely, the faulty ones have at most 1/3 of the total voting power).
 
-When we say we trust _h.Header.NextV_ we do _not_ trust that each individual validator in _h.Header.NextV_ is correct, but we only trust the fact that at most 1/3 of them are faulty (more precisely, the faulty ones have at most 1/3 of the total voting power).
 
 ### Functions
 
-The function _Bisection_ checks whether to trust header _h2_ based on the trusted header _h1_. It does so by calling
-the function _CheckSupport_ in the process of
-bisection/recursion. _CheckSupport_ implements the trusted period method and, for two adjacent headers (in term of heights), it checks uninterrupted sequence of proof.
+The function *CanTrust* checks whether to trust header *h2* based on the trusted header *h1*. It does so by (potentially)
+building transitive trust relation between *h1* and *h2*, over some intermediate headers. For example, in case we cannot trust
+header *h2* based on the trusted header *h1*, the function *CanTrust* will try to find headers such that we can transition trust
+from *h1* over intermediate headers to *h2*. We will give two implementations of *CanTrust*, the one based
+on bisection that is recursive and the other that is non-recursive. We give two implementations as recursive version might be easier
+to understand but non-recursive version might be simpler to formally express and verify using TLA+/TLC.
 
-_Assumption_: In the following, we assume that _h2.Header.height > h1.Header.height_. We will quickly discuss the other case in the next section.
+Both implementations of *CanTrust* function are based on *CheckSupport* function that implements the skipping conditions under which we can trust a
+header *h2* given the trust in the header *h1* as a single step,
+i.e., it does not assume ensuring transitive trust relation between headers through some intermediate headers.
 
-We consider the following set-up:
+In order to incentivize correct behavior of validators that run Tendermint consensus protocol, fork detection protocol (it will be explained in different document) is executed in case of a fork (conflicting
+headers are detected). As detecting conflicting headers, its propagation through the network (by the gossip protocol) and execution of the fork accountability
+protocol on the chain takes time, the lite client logic assumes conservative value for trusted period. More precisely, in the context of lite client we always
+operate with a smaller trusted period that we call *lite client trusted period* (LITE_CLIENT_TRUSTED_PERIOD). If we assume that upper bound
```

Comment:
This is a bit tricky. My understanding is that trust model from Tendermint Failure Model defines period during which processes behave correctly, i.e., UNBONDING_PERIOD. You can think about this as normal case, similar as we assume in Tendermint that <1/3 are faulty. On the other side in cases Tendermint Failure Model is violated we want to be able to process detected violation (fork) on time, i.e., within UNBONDING_PERIOD. Therefore we define smaller trusted period for the lite client(LITE_CLIENT_TRUSTED_PERIOD). In a sense, this logic is taking into account possibility of Tendermint failure model being violated.

Reply:
I agree that we need two periods - one for "unbonding" and one for "trusting". But my contention is that the Failure Model only needs to be defined in the "trusting" one - it doesn't care about the longer "unbonding" period, because that extra time is just for us to detect failure and submit evidence, ie where the correctness assumption is already violated! Hence the light client doc just needs one period for both the failure model and the light client protocol, but an actually implementation would need a second longer period during which misbehaviour could be punished on chain (though note this is application specific, not everyone will use the same concept of a fixed unbonding period, maybe some of your stake is slashable until the end of time).

Reply:
I see your point. I think this was also Josef's original intention. I find confusing using that one period to define Tendermint failure model. So maybe we need to move Tendermint failure model to some other spec and then lite client is defined with a single period. Or you think that it's fine just leaving it as was before? The reason I suggested this change is the part when we reason about header expiration. I felt that we need to make distinction between real header expiration (UNBONDING_PERIOD related) vs lite client expiration (smaller period). @josef-widder Can you explain what was your original intention and what is your view on this discussion?

Reply:
My understanding is that we need the failure model to define when we can assume some processes to behave correctly. This is defined by the TRUSTED_PERIOD. I do not think that we can assume processes to behave correctly during the whole UNBONDING_PERIOD, as we have to incorporate the detection and punishing. If we need a day for detecting and punishing, the unbonding period should end a day after the trusting period. I think this is independent of whether the 1/3 assumption is violated. 

Thus, to *define* the failure model, we should use the TRUSTED_PERIOD. To *enforce* the failure model (via accountability, detection, punishing) we need the UNBONDING_PERIOD.

I agree that the lite client spec is not the right place to define the Tendermint failure model. I put it there, because we needed it, and it wasn't written explicitly anywhere else at that time. At least that was my impression.

Assuming we can bound the detection and punishing time, from the unbonding period, we will get a maximal TRUSTED_PERIOD that we can make explicit as Tendermint failure model. Any protocol that wants to communicate with Tendermint can of course be designed with a shorter trusted period. This means it works under less strict fault assumptions than provided by Tendermint (that is, by the detection and punishing scheme).

Reply:
Also, to be clear, I think this failure model is defined for "outside" observers only. For consensus, my current understanding is that we assume that more than 2/3 of the validators are correct until the current consensus instance terminates; no periods here. That is, we have different failure assumptions for different protocols. The important thing is to make explicit for each protocol what is assumed. Then we will have to check/make sure that the combination of the fault assumptions is consistent.

Reply:
Sounds spot on. Given that final comment (different models for different protocols), I think a failure model does belong in this document, specifically the Tendermint Light Client Failure Model. A separate model exists for the full node (Tendermint Full Node Failure Model).

Then for the Light Client Failure Model, we only need the LIGHT_CLIENT_TRUSTING_PERIOD, which can be just called TRUSTING_PERIOD. The unbonding period should be part of a fork-detection/accountability model and tracked in one of those documents.